compiler/src/iree/compiler/Dialect/Modules/VMVX/Transforms/DecomposeLinalgGeneric.cpp

```diff
@@ -0,0 +1,220 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Modules/VMVX/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Modules/VMVX/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassRegistry.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace VMVX {
+
+namespace {
+
+struct DecomposeParallelGeneric
+    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+  struct BodyValueInfo {
+    // The generic operand that this body value maps to.
+    Value outerValue;
+    // The indexing map associated with this body value.
+    AffineMap indexingMap;
+  };
+  struct OpInfo {
+    linalg::LinalgOp op;
+
+    // The iteration domain, either as statically known values or dynamically
+    // known symbols. This will equal the number of loops and the number of
+    // dimensions in each indexing map.
+    SmallVector<OpFoldResult> domain;
+
+    // All dimensions of the domain, using ShapedType::kDynamicDim to signify
+    // a dynamic dimension.
+    SmallVector<int64_t> domainTypeDims;
+
+    // Values for the subset of dynamic dims in domainTypeDims;
+    SmallVector<Value> domainTypeDynamicDims;
+
+    // Map of body SSA values to indexing map/outer value. If the body value
+    // is not associated with a generic operand (i.e. it is captured), it will
+    // not be in this map.
+    llvm::DenseMap<Value, BodyValueInfo> bodyValueInfo;
+
+    // Iterator types as a vector for easier construction of ops.
+    SmallVector<StringRef> iteratorTypes;
+
+    OpInfo(linalg::LinalgOp &op) : op(op) {
+      for (Attribute attr : op.getIteratorTypes()) {
+        iteratorTypes.push_back(attr.cast<StringAttr>().getValue());
+      }
+    }
+
+    Value createDomainInitTensor(Type elementType, OpBuilder &builder) {
+      return builder.create<linalg::InitTensorOp>(
+          op.getLoc(), domainTypeDynamicDims, domainTypeDims, elementType);
+    }
+  };
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp op,
```

Comment:
First read of this, I think this works fine. It definitely has the advantage of lesser compilation time. You create exactly the generic ops you need.

Reply:
Thanks for the look. Let's have a look in the context of the whole PR (which now has a bunch more) and decide what to do with it. I mainly was wondering if I was generally on an ok path.

Reply:
Since there is a bunch of stuff going on with the PR, if you think it makes sense, I can take a stab at quickly adding this pattern upstream. We can then pull that into this PR and see if it does the job. Mainly interested in stress testing the canonicalization pattern that should work in tandem with the pattern I had in mind, and apart from lit tests havent had a chance to stress test it.

Reply:
I'm not going to say no to the help :)

Realistically, this will be Friday before I am ready to land anything here. I'm just flushing out full-system bugs and haven't even done the real part or finishing work yet.

Reply:
Ok, I think I can get that worked out soon. Will let you know when I have something.